sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java

```diff
@@ -0,0 +1,562 @@
+
+      input.apply(
+        FileIO.<GenericRecord>write()
+          .via(sink(getSchema()))
+          .to(prefix.getCurrentDirectory().toString())
+          .withPrefix(prefix.getFilename())
+          .withSuffix(".parquet"));
+
+      return PDone.in(input.getPipeline());
+    }
+  }
+
+  /**
+   * Creates a {@link Sink} that, for use with {@link FileIO#write}.
+   */
+  public static Sink<GenericRecord> sink(Schema schema) {
+    return new AutoValue_ParquetIO_Sink.Builder<GenericRecord>()
+      .setJsonSchema(schema.toString())
+      .build();
+  }
+
+  /** Implementation of {@link #sink}. */
+  @AutoValue
+  public abstract static class Sink<T> implements FileIO.Sink<GenericRecord> {
+
+    abstract String getJsonSchema();
+
+    abstract Builder<T> toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder<T> {
+      abstract Builder<T> setJsonSchema(String jsonSchema);
+      abstract Sink<T> build();
+    }
+
+    @Nullable private transient ParquetWriter<GenericRecord> writer;
+
+    @Nullable private transient Schema schema;
+
+    @Override
+    public void open(WritableByteChannel channel) throws IOException {
+      this.schema = new Schema.Parser().parse(getJsonSchema());
+
+      BeamParquetOutputFile beamParquetOutputFile =
+        new BeamParquetOutputFile(Channels.newOutputStream(channel));
+
+      this.writer = AvroParquetWriter.<GenericRecord>builder(beamParquetOutputFile)
+        .withSchema(schema)
+        .withWriteMode(OVERWRITE)
+        .build();
+    }
+
+    @Override
+    public void write(GenericRecord element) throws IOException {
+      checkNotNull(writer, "Writer cannot be null");
+      writer.write(element);
+    }
+
+    @Override
+    public void flush() throws IOException {
+      writer.close();
+    }
+
+    private static class BeamParquetOutputFile implements OutputFile {
+
+      private OutputStream outputStream;
+
+      BeamParquetOutputFile(OutputStream outputStream) {
+        this.outputStream = outputStream;
+      }
+
+      @Override public PositionOutputStream create(long blockSizeHint) throws IOException {
+        return new BeamOutputStream(outputStream);
+      }
+
+      @Override public PositionOutputStream createOrOverwrite(long blockSizeHint) {
+        return new BeamOutputStream(outputStream);
+      }
+
+      @Override public boolean supportsBlockSize() {
+        return false;
+      }
+
+      @Override public long defaultBlockSize() {
+        return 0;
+      }
+    }
+
+    private static class BeamOutputStream extends PositionOutputStream {
+      private long position = 0;
+      private OutputStream outputStream;
+
+      private BeamOutputStream(OutputStream outputStream) {
+        this.outputStream = outputStream;
+      }
+
+      @Override public long getPos() throws IOException {
+        return position;
+      }
+
+      @Override public void write(int b) throws IOException {
```

Comment:
We need to provide overrides for writing byte array slices as well. Otherwise all writing will happen byte-by-byte, which will be insanely slow.